de.metas.adempiere.libero.libero/src/main/java/org/eevolution/process/PP_Product_BOM_Check.java

```diff
@@ -18,56 +18,142 @@
  *
  * @author Tony Snook (tspc)
  * @author Teo Sarca, SC ARHIPAC SERVICE SRL
  */
-public class PP_Product_BOM_Check extends JavaProcess
+public class PP_Product_BOM_Check extends JavaProcess implements IProcessPrecondition
 {
 	private final transient IProductBOMBL productBOMBL = Services.get(IProductBOMBL.class);
 	private final transient IProductBOMDAO productBOMDAO = Services.get(IProductBOMDAO.class);
 	private final transient ITrxManager trxManager = Services.get(ITrxManager.class);
 
+	@Param(parameterName = I_M_Product.COLUMNNAME_M_Product_Category_ID, mandatory = false)
+	private int p_M_Product_Category_ID;
+
+	@Override
+	public ProcessPreconditionsResolution checkPreconditionsApplicable(IProcessPreconditionsContext context)
+	{
+		if (!(I_M_Product.Table_Name.equals(context.getTableName()) || I_PP_Product_BOM.Table_Name.equals(context.getTableName())))
+		{
+			return ProcessPreconditionsResolution.reject();
+		}
+
+		if (context.getSelectionSize()>1)
+		{
+			return ProcessPreconditionsResolution.reject();
+		}
+
+		return ProcessPreconditionsResolution.accept();
+	}
+
+
 	@Override
 	@RunOutOfTrx
 	protected String doIt()
 	{
 		log.info("Check BOM Structure");
 
+		if (p_M_Product_Category_ID > 0)
+		{
+			final IQueryBuilder<I_M_Product> queryBuilder = Services.get(IQueryBL.class)
+					.createQueryBuilder(I_M_Product.class, getCtx(), get_TrxName())
+					.addEqualsFilter(I_M_Product.COLUMNNAME_IsBOM, true);
+
+			if (p_M_Product_Category_ID > 0)
+			{
+				queryBuilder.addEqualsFilter(I_M_Product.COLUMNNAME_M_Product_Category_ID, p_M_Product_Category_ID);
+			}
+
+			queryBuilder.orderBy()
+					.addColumn(I_M_Product.COLUMNNAME_Name)
+					.endOrderBy();
+
+			final AtomicInteger counter = new AtomicInteger(0);
+
+			queryBuilder.create()
+					.stream()
+					.forEach(product -> {
+
+						final boolean isvalidated = validateProduct(product);
+						if (isvalidated)
+						{
+							counter.incrementAndGet();
+						}
+					});
+
+			return "#" + counter.get();
+		}
+		else
+		{
+			final I_M_Product product = InterfaceWrapperHelper.create(getCtx(), getM_Product_ID(), I_M_Product.class, get_TrxName());
+			validateProduct(product);
+			return MSG_OK;
+		}
+	}
+
+
+	private int getM_Product_ID()
+	{
+		final String tableName = getTableName();
+		if (I_M_Product.Table_Name.equals(tableName))
+		{
+			return getRecord_ID();
+		}
+		else if (I_PP_Product_BOM.Table_Name.equals(tableName))
+		{
+			final I_PP_Product_BOM bom = getRecord(I_PP_Product_BOM.class);
+			return bom.getM_Product_ID();
+		}
+		else
+		{
+			throw new AdempiereException(StringUtils.formatMessage("Table {} has not yet been implemented to support BOM validation.", tableName));
+		}
+
+	}
+
+	private boolean validateProduct(@NonNull final I_M_Product product)
+	{
 		try
 		{
-			trxManager.run(() -> checkProductById(getRecord_ID()));
+			trxManager.run(() -> checkProductById(product));
+			return true;
 		}
 		catch (final Exception ex)
```

Comment:
log.warn the exception and product. We shall be able to find out why it failed.
More, why not having a second param, called throwEx and in case it's true we shall proparate the exception.
In case of single product validation (i.e. when productCategorId <= 0) we shall propagate the exception directly.